src/Compilers/Core/Portable/Emit/EditAndContinue/DeltaMetadataWriter.cs

```diff
@@ -338,14 +338,24 @@ protected override int GetParameterDefIndex(IMethodDefinition methodDef)
             return 0;
         }
 
-        protected override int GetOrAddAssemblyRefIndex(IAssemblyReference reference)
+        protected override int GetOrAddAssemblyRefIndex(AssemblyIdentity reference)
         {
             return _assemblyRefIndex.GetOrAdd(reference);
         }
 
-        protected override IReadOnlyList<IAssemblyReference> GetAssemblyRefs()
+        protected override IReadOnlyList<AssemblyIdentity> GetAssemblyRefs()
         {
-            return _assemblyRefIndex.Rows;
+            if (_previousGeneration.InitialBaseline.LazyMetadataSymbols.AssemblyReferenceIdentityMap.IsEmpty)
+            {
+                return _assemblyRefIndex.Rows;
+            }
+
+            // only happens in presence of wildcard versions:
+            return _assemblyRefIndex.Rows.Select(identity =>
+            {
+                AssemblyIdentity mapped;
+                return _previousGeneration.InitialBaseline.LazyMetadataSymbols.AssemblyReferenceIdentityMap.TryGetValue(identity, out mapped) ? mapped : identity;
+            }).Distinct().ToArray();
```

Comment:
Wouldn't Distinct make some previously allocated assembly ref indexes invalid if it removes something? I would feel that _assemblyRefIndex should be populated in such a way that we don't get any duplicates here.